src/arithmetic/algebraic_expression/algebraic_expression_optimization.c

```diff
@@ -411,6 +413,109 @@ static void _AlgebraicExpression_PushDownTranspose(AlgebraicExpression *root) {
+// Replace transpose ops with transposed operands.
+//------------------------------------------------------------------------------
+// Transpose an operand matrix and update the expression accordingly.
+static void _AlgebraicExpression_TransposeOperand(AlgebraicExpression *operand) {
+	// Swap the row and column domains of the operand.
+	const char *tmp = operand->operand.dest;
+	operand->operand.src = operand->operand.dest;
+	operand->operand.dest = tmp;
+
+	// Diagonal matrices do not need to be transposed.
+	if(operand->operand.diagonal == true) return;
+
+	GrB_Type type;
+	GrB_Index nrows;
+	GrB_Index ncols;
+	GrB_Matrix replacement;
+	// Create a new empty matrix with the type and dimensions of the original.
+	GrB_Matrix_nrows(&nrows, operand->operand.matrix);
+	GrB_Matrix_ncols(&ncols, operand->operand.matrix);
+	GxB_Matrix_type(&type, operand->operand.matrix);
+	GrB_Info info = GrB_Matrix_new(&replacement, type, nrows, ncols);
+	if(info != GrB_SUCCESS) {
+		fprintf(stderr, "%s", GrB_error());
+		assert(false);
+	}
+
+	// Populate the replacement with the transposed contents of the original.
+	info = GrB_transpose(replacement, GrB_NULL, GrB_NULL, operand->operand.matrix, GrB_NULL);
+	if(info != GrB_SUCCESS) {
+		fprintf(stderr, "%s", GrB_error());
+		assert(false);
+	}
+
+	// Update the matrix pointer.
+	operand->operand.matrix = replacement;
+	// As this matrix was constructed, it must ultimately be freed.
+	operand->operand.bfree = true;
+}
+
+/* Find transpose operations with an operand child and replace them with an actual transposed
+ * operand. This optimization should be valid for all transpose operations at this point.
+ *
+ * If at this point we have the tree:
+ *
+ *                (+)
+ *    (transpose)     (transpose)
+ *         (A)            (B)
+ *
+ * It will become:
+ *                (+)
+ *           (A')     (B')
+ *
+ * Similarly, if we have:
+ *
+ *                (*)
+ *    (transpose)     (transpose)
+ *         (B)            (A)
+ *
+ * It will become:
+ *                (*)
+ *           (B')     (A')
+ */
+static void _AlgebraicExpression_ApplyTranspose(AlgebraicExpression *root) {
+	uint child_count = 0;
+	AlgebraicExpression *child = NULL;
+
+	switch(root->type) {
+	case AL_OPERAND:
+		break;  // Nothing to be done.
+
+	case AL_OPERATION:
+		switch(root->operation.op) {
+		case AL_EXP_ADD:    // Fall through.
+		case AL_EXP_MUL:    // Fall through.
+		case AL_EXP_POW:    // Fall through.
+			child_count = AlgebraicExpression_ChildCount(root);
+			for(uint i = 0; i < child_count; i++) {
+				AlgebraicExpression *child = root->operation.children[i];
+				_AlgebraicExpression_ApplyTranspose(child);
+			}
+			break;
+
+		case AL_EXP_TRANSPOSE:
+			// Transpose operands will currently always have an operand child, but
+			// since evalution can perform transposition we don't need to assert here.
+			if(root->operation.children[0]->type != AL_OPERAND) break;
```

Comment:
keep calling `_AlgebraicExpression_ApplyTranspose` recursively in the same manner as all other operations,
consider: `T(T(A) * T(B))`

Reply:
After applying `_AlgebraicExpression_PushDownTranspose`, there are no transpose operations that do not have a single operand child. Your example at this point would be `A * T(B)`.

Do you want this change made for consistency regardless?

Reply:
Yes please.